ament_index_cpp/include/ament_index_cpp/get_package_prefix.hpp

```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef AMENT_INDEX_CPP__GET_PACKAGE_PREFIX_HPP_
+#define AMENT_INDEX_CPP__GET_PACKAGE_PREFIX_HPP_
+
+#include <stdexcept>
+#include <string>
+
+#include "ament_index_cpp/visibility_control.h"
+
+namespace ament_index_cpp
+{
+
+/// Thrown when a package is not found.
+class PackageNotFoundError : public std::out_of_range
+{
+public:
+  AMENT_INDEX_CPP_PUBLIC
+  explicit PackageNotFoundError(const std::string & package_name);
+
+  AMENT_INDEX_CPP_PUBLIC
+  virtual ~PackageNotFoundError();
+
+  const std::string package_name;
```

Comment:
Why should the member be `const`? For a non-const instance of this class the user might want to change this member.

Reply:
Because the package name is baked into the error message, and if you change this then it would be inconsistent with the `what()`, so you should only set it from the constructor. If you want a different package name in the error, I'd say create another object.

Reply:
I still don't think that the class should prescribe the member to be `const`.

Reply:
Unfortunately I don't agree. I've given a reason why it would be undesirable to change the package name after creation, but I haven't seen a reason why you would want to change it. I'm willing to change it if a convincing reason was given and that outweighs the downside I've already stated.

Reply:
The package for which the exception was thrown does not seem like it should be modifiable after the fact or in the traceback/error handling thus keeping it const would make sense. And relaxing the const constraint in the future will always be easier than adding it if there is a use case that needs it than the reverse.

Reply:
Btw, I looked at making a setter and getter for the `package_name`, as a compromise, but that would require me to override the `what()` method (storage for `what()` is not exposed in the exception class) and provide my own storage which would be redundant. Also the `what()` function is `noexcept`:

http://en.cppreference.com/w/cpp/error/exception/what

So, the creating of the new `std::string` storage would need to be done in the setter, more or less what the constructor is doing. So I don't think it would be any easier (more efficient) than creating a new exception if you needed to change the package name (still don't know why you would need to do so).